### PR TITLE
refactor(scan): implement CI discovery provider

### LINKED ISSUE #88: Implement CI discovery provider
Implement CI Discovery provider compatible with [ci-info](https://github.com/watson/ci-info) package which is used by sectester-js

### PR DESCRIPTION
**Pull request by ostridm**
closes #88

### CODE REVIEW

File: src/SecTester.Scan/CI/DefaultCiDiscovery.cs [hunk 1]
```diff
@@ -1,9 +1,48 @@
...
+    {
+      return;
+    }
+
+    Server = typeof(CiServer)
+      .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+      .Where(x => x.Name.Equals(vendor.Constant, StringComparison.OrdinalIgnoreCase))
+      .Select(x => x.GetValue(null))
+      .Cast<CiServer>()
+      .FirstOrDefault() ?? new CiServer(vendor.Name);
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/DefaultCiDiscovery.cs]: `CiServer` is an enumeration, why don't move that logic to it? (e.g. `FromValue()` or `FromName`)

File: src/SecTester.Scan/CI/CiServer.cs [hunk 2]
```diff
@@ -2,5 +2,50 @@ namespace SecTester.Scan.CI;
 
 public record CiServer(string? ServerName = default)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: It is hardly a record, covert to class and implement a `IComparable` interface

File: src/SecTester.Scan/CI/CiServer.cs [hunk 3]
```diff
@@ -2,5 +2,50 @@ namespace SecTester.Scan.CI;
...
 {
-  public static readonly CiServer? Unknown;
+  public static CiServer APPCENTER { get; } = new("Visual Studio App Center");
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Please follow the naming convention, instead of using a property name introduce an `id` or `value` field. Use it to filter enumeration.

```suggestion
  public static CiServer AppCenter { get; } = new("Visual Studio App Center");
```

File: src/SecTester.Scan/CI/DefaultCiDiscovery.cs [hunk 4]
```diff
@@ -1,9 +1,48 @@
...
   public bool IsCi => Server != null;
+
+  public bool IsPr { get; }
+
+  public DefaultCiDiscovery(IDictionary? env = default)
+  {
+    env ??= Environment.GetEnvironmentVariables();
+
+    var vendors = JsonSerializer.Deserialize<Vendor[]>(
+      ResourceUtils.GetEmbeddedResourceContent<DefaultCiDiscovery>("SecTester.Scan.CI.vendors.json"),
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/DefaultCiDiscovery.cs]: Extract a string as constant

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  private bool MatchOwnProperties(JsonElement element)
+  {
+    return element.EnumerateObject().All(x =>
+      x.Value.ValueKind == JsonValueKind.String && _env.Contains(x.Name) &&
+      x.Value.ValueEquals(_env[x.Name]?.ToString()));
+  }
+
+  private bool MatchInnerAny(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  private bool Any(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  {
+    var any = element.EnumerateObject().FirstOrDefault(x => x.NameEquals("any"));
+
+    return any.Value.ValueKind == JsonValueKind.Array && any.Value.EnumerateArray()
+      .Any(x => x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!));
+  }
+
+  private bool MatchInnerEnvIncludes(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  private bool Includes(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+    if (env is null || includes is null || !_env.Contains(env))
+    {
+      return false;
+    }
+
+    var envVarValue = _env[env]?.ToString();
+    return envVarValue?.Contains(includes) ?? false;
+  }
+
+  private bool MatchInnerEnvNe(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  private bool NotEqual(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+  {
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => element.GetString() is not null && _env.Contains(element.GetString()!),
+      JsonValueKind.Object => MatchInnerEnvNe(element) || MatchEnvElement(element),
+      _ => false
+    };
+  }
+
+  private bool MatchOwnProperties(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  private bool HasOwnProperties(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+internal class VendorMatcher
+{
+  private readonly IDictionary _env;
+
+  public VendorMatcher(IDictionary env)
+  {
+    _env = env;
+  }
+
+  public bool MatchEnvElement(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  public bool MatchEnv(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+      JsonValueKind.Object => MatchInnerAny(element) || MatchInnerEnvIncludes(element) ||
+                              MatchOwnProperties(element),
+      JsonValueKind.Array => element.EnumerateArray()
+        .All(x => x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)),
+      _ => false
+    };
+  }
+
+
+  public bool MatchPrElement(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  public bool MatchPr(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    if (env is null || ne is null || !_env.Contains(env) || ne != "false")
+    {
+      return false;
+    }
+
+    var envVarValue = _env[env]?.ToString();
+    return !string.IsNullOrEmpty(envVarValue);
+  }
+
+  private static string? GetInnerPropertyValue(JsonElement element, string name)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]:
```suggestion
  private static string? GetPropertyValue(JsonElement element, string name)
```

File: test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    PropertyNameCaseInsensitive = true, PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+  };
+
+  public static readonly IEnumerable<object[]> MatchEnvInput = new List<object[]>
+  {
+    new object[]
+    {
+      @"{
+        ""env"": { ""env"": ""NODE"", ""includes"": ""/app/.heroku/node/bin/node""}
+        }",
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs]: Avoid the wrong indentation
```suggestion
      @"{""env"": { ""env"": ""NODE"", ""includes"": ""/app/.heroku/node/bin/node""}}",
```

File: test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs [hunk 13]
```diff
@@ -0,0 +1,252 @@
+using System.Collections.Specialized;
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs]: global using

File: test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    var vendors = JsonSerializer.Deserialize<Vendor[]>(
+      ResourceUtils.GetEmbeddedResourceContent<DefaultCiDiscovery>("SecTester.Scan.CI.vendors.json"),
+      new JsonSerializerOptions() { PropertyNameCaseInsensitive = true });
+
+    // act
+    var result = vendors!.FirstOrDefault(x=> sut.MatchPrElement(x.Pr));
+    
+    // assert
+    result.Should().BeNull();
+  }
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs]: Those tests do not test anything, they just overlap the tests above

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+  public VendorMatcher(IDictionary env)
+  {
+    _env = env;
+  }
+
+  public bool MatchEnvElement(JsonElement element)
+  {
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => element.GetString() is not null && _env.Contains(element.GetString()!),
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: It may lead to FP in the case of an empty string. Also, you have to pass `StringComparison` second argument

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+      JsonValueKind.Object => MatchInnerAny(element) || MatchInnerEnvIncludes(element) ||
+                              MatchOwnProperties(element),
+      JsonValueKind.Array => element.EnumerateArray()
+        .All(x => x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)),
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: Duplicate `MatchInnerAny`

File: src/SecTester.Scan/CI/VendorMatcher.cs (hunk 15 above)
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: Also duplicated multiple times

**Review by derevnjuk — CHANGES_REQUESTED** [review]
Look fine, but you still have what needs to be improved and refactored

File: src/SecTester.Scan/CI/VendorMatcher.cs (hunk 16 above)
Comment (ostridm) [reply src/SecTester.Scan/CI/VendorMatcher.cs]: `All` vs `Any`

Comment (derevnjuk) [reply src/SecTester.Scan/CI/VendorMatcher.cs]: What difference that makes?

```diff
- x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)
+ x.ValueKind == JsonValueKind.String && x.GetString() is not null && _env.Contains(x.GetString()!)
```

File: src/SecTester.Scan/CI/CiServer.cs [hunk 17]
```diff
@@ -1,6 +1,125 @@
...
+  public static CiServer Vercel { get; } = new("VERCEL", "Vercel");
+  public static CiServer Woodpecker { get; } = new("WOODPECKER", "Woodpecker");
+  public static CiServer XcodeCloud { get; } = new("XCODE_CLOUD", "Xcode Cloud");
+  public static CiServer XcodeServer { get; } = new("XCODE_SERVER", "Xcode Server");
+
+  private readonly string _constant;
+  private readonly string _name;
+  private readonly int _hashcode;
+
+  public string Constant => _constant;
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: `Constant`? What does it mean? You must utilize clear and concise names, e.g. `Name` and `DisplayName`, or `Id` and `Name`.

File: src/SecTester.Scan/CI/CiServer.cs [hunk 18]
```diff
@@ -1,6 +1,125 @@
...
+  }
+
+  public bool Equals(CiServer? other)
+  {
+    if (other is null)
+    {
+      return false;
+    }
+
+    return ReferenceEquals(Constant, other.Constant) || Constant.Equals(other.Constant, StringComparison.OrdinalIgnoreCase);
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: It is not clear why you need to do the same by `ReferenceEquals`

```suggestion
    return Constant.Equals(other.Constant, StringComparison.OrdinalIgnoreCase);
```

File: src/SecTester.Scan/CI/CiServer.cs [hunk 19]
```diff
@@ -1,6 +1,125 @@
...
+
+  public override int GetHashCode()
+  {
+    return _hashcode;
+  }
+
+  public override string ToString()
+  {
+    return _constant;
+  }
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: It is not clear to me why don't show humanized branded name

```suggestion
  public override string ToString() => _name;
```

File: src/SecTester.Scan/CI/CiServer.cs [hunk 20]
```diff
@@ -1,6 +1,125 @@
...
+  {
+    _constant = constant;
+    _name = name;
+    _hashcode = StringComparer.OrdinalIgnoreCase.GetHashCode(_constant);
+  }
+
+  public override int GetHashCode()
+  {
+    return _hashcode;
+  }
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]:
```suggestion
  public override int GetHashCode() => _hashcode;
```

File: src/SecTester.Scan/CI/CiServer.cs [hunk 21]
```diff
@@ -1,6 +1,125 @@
...
+  public static bool operator !=(CiServer? left, CiServer? right)
+  {
+    return !(left == right);
+  }
+
+  internal static CiServer From(Vendor vendor)
+  {
+    return From(vendor.Constant, vendor.Name);
+  }
+  public static CiServer From(string constant, string name)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]:
```suggestion
  }
  
  public static CiServer From(string constant, string name)
```

Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Well, quite a strange method 🤦🏾  In your interpretation this is a static factory method, it seems you love to utilize patterns randomly)

File: src/SecTester.Scan/CI/CiServer.cs [hunk 22]
```diff
@@ -1,6 +1,125 @@
...
+  {
+    if (string.IsNullOrEmpty(constant))
+    {
+      throw new ArgumentException("Constant value must not be empty.");
+    }
+
+    return typeof(CiServer)
+             .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+             .Select(x => x.GetValue(null))
+             .Cast<CiServer>()
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Extract to a separate method (e.g.  `GetAll`)

File: src/SecTester.Scan/CI/CiServer.cs [hunk 23]
```diff
@@ -1,6 +1,125 @@
...
+             .FirstOrDefault(x => x.Constant.Equals(constant, StringComparison.OrdinalIgnoreCase))
+           ?? new CiServer(constant, name);
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Return null instead, deciding what to do is the responsibility of the caller code

File: src/SecTester.Scan/CI/CiServer.cs [hunk 24]
```diff
@@ -1,6 +1,125 @@
...
+  public static CiServer Woodpecker { get; } = new("WOODPECKER", "Woodpecker");
+  public static CiServer XcodeCloud { get; } = new("XCODE_CLOUD", "Xcode Cloud");
+  public static CiServer XcodeServer { get; } = new("XCODE_SERVER", "Xcode Server");
+
+  private readonly string _constant;
+  private readonly string _name;
+  private readonly int _hashcode;
+
+  public string Constant => _constant;
+  public string Name => _name;
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Why don't convert it to auto-prop?

File: src/SecTester.Scan/CI/CiServer.cs (hunk 17 above)
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: ditto

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+  {
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => CheckStringValue(element),
+      JsonValueKind.Object => NotEqual(element) || MatchEnv(element),
+      _ => false
+    };
+  }
+
+  private bool CheckStringValue(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: Unclear and superficial name, it is intended to check what? (e.g. `HasProperty`)

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+      _ => false
+    };
+  }
+
+  public bool MatchPr(JsonElement element)
+  {
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => CheckStringValue(element),
+      JsonValueKind.Object => NotEqual(element) || MatchEnv(element),
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: What is the point to call the whole `MatchEnv`? Extract `JsonValueKind.Object => Any(element) || Includes(element) || HasOwnProperties(element)` to a separate method

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+    return element.ValueKind switch
+    {
+      JsonValueKind.String => CheckStringValue(element),
+      JsonValueKind.Object => Any(element) || Includes(element) || HasOwnProperties(element),
+      JsonValueKind.Array => element.EnumerateArray().All(CheckStringValue),
+      _ => false
+    };
+  }
+
+  public bool MatchPr(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs:28]: The only difference between `MatchEnv` and `MatchPr` is in using `NotEqual`. Why don't extract a common logic to a separate method

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 28]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  private bool Any(JsonElement element)
+  {
+    var any = element.EnumerateObject().FirstOrDefault(x => x.NameEquals("any"));
+
+    return any.Value.ValueKind == JsonValueKind.Array && any.Value.EnumerateArray().Any(CheckStringValue);
+  }
+
+  private bool Includes(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs:62]: `Includes` and `NotEqual` are identical, the difference only in the last statement
```csharp
return envVarValue?.Contains(includes) ?? false;
// vs
return ne == "false" && !string.IsNullOrEmpty(envVarValue);
```
or even better
```csharp
return envVarValue?.Contains(includes) ?? false;
// vs
return !envVarValue?.Equals(ne) ?? false;
```

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/SecTester.Scan/CI/CiServer.cs [hunk 29]
```diff
@@ -1,6 +1,107 @@
...
+  public static CiServer Travis { get; } = new("TRAVIS", "Travis CI");
+  public static CiServer Vercel { get; } = new("VERCEL", "Vercel");
+  public static CiServer Woodpecker { get; } = new("WOODPECKER", "Woodpecker");
+  public static CiServer XcodeCloud { get; } = new("XCODE_CLOUD", "Xcode Cloud");
+  public static CiServer XcodeServer { get; } = new("XCODE_SERVER", "Xcode Server");
+
+  private static readonly IEnumerable<CiServer> All = typeof(CiServer)
+    .GetProperties(BindingFlags.Public | BindingFlags.Static | BindingFlags.DeclaredOnly)
+    .Select(x => x.GetValue(null))
+    .Cast<CiServer>();
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Avoid any redundant optimization made without any purpose, especially when it was not requested. Doing any useless changes after review, which have not been requested, just delays the merge. Move to the method.

File: src/SecTester.Scan/CI/CiServer.cs [hunk 30]
```diff
@@ -1,6 +1,107 @@
...
+  private readonly int _hashcode;
+
+  public string Id { get; }
+  public string Name { get; }
+
+  public CiServer(string id, string name)
+  {
+    if (string.IsNullOrEmpty(id))
+    {
+      throw new ArgumentException("Id must not be empty.");
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]: Can it accept null? Clarify the message or use `ArgumentNullException`

```suggestion
      throw new ArgumentException("ID must not be empty.");
```

File: src/SecTester.Scan/CI/CiServer.cs [hunk 31]
```diff
@@ -1,6 +1,107 @@
...
+  }
+
+  public override int GetHashCode() => _hashcode;
+
+  public override string ToString() => Name;
+
+  public override bool Equals(object? obj)
+  {
+    return Equals(obj as CiServer);
+  }
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/CiServer.cs]:
```suggestion
  public override bool Equals(object? obj) => Equals(obj as CiServer)
```

File: src/SecTester.Scan/CI/DefaultCiDiscovery.cs [hunk 32]
```diff
@@ -1,9 +1,44 @@
...
+
+    var vendor = vendors.FirstOrDefault(x => matcher.MatchEnv(x.Env));
+
+    if (vendor is null)
+    {
+      return;
+    }
+
+    Server = CiServer.GetAll()
+               .FirstOrDefault(x => vendor.Constant.Equals(x.Id, StringComparison.OrdinalIgnoreCase))
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/DefaultCiDiscovery.cs]: Why did you remove `FromId` (aka `From`)? If you would be a hero, your name was "Duplicatemen" 😄  You make me repeat comments multiple times https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045193362

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+      _ => false
+    };
+  }
+
+  private bool MatchObject(JsonElement element)
+  {
+    return Any(element) || Includes(element) || HasOwnProperties(element);
+  }
+
+  private bool HasValue(JsonElement element)
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs]: `Contains` just verifies if there is a property, but cannot say anything about value https://github.com/NeuraLegion/sectester-net/pull/94#discussion_r1045195231

```suggestion
  private bool HasProperty(JsonElement element)
```

File: src/SecTester.Scan/CI/VendorMatcher.cs [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+  private bool NotEqual(JsonElement element)
+  {
+    return ApplyPredicate(element, "ne", (envVarValue, neValue) =>
+      neValue == "false" && !string.IsNullOrEmpty(envVarValue)
+    );
+  }
+
+  private bool ApplyPredicate(JsonElement element, string propertyName, Func<string?, string, bool> predicate)
+  {
+    var env = GetPropertyValue(element, "env");
```
Comment (derevnjuk) [inline src/SecTester.Scan/CI/VendorMatcher.cs:78]: The method is too complex. Extract a logic to retrieve env to a separate method.

```csharp
private string? GetEnvValue(JsonElement element)
{
  var env = GetPropertyValue(element, "env");

  if (string.IsNullOrEmpty(env) || !_environment.Contains(env))
  {
    return null;
  }

  return _environment[env]?.ToString();
}
```

File: test/SecTester.Scan.Tests/CI/CiServerTests.cs [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+    {
+      null!, ""
+    },
+    new object[]
+    {
+      string.Empty, ""
+    }
+  };
+
+  public static IEnumerable<object[]> EqualityInput = new List<object[]>
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/CiServerTests.cs]: OMG... Never use functions to generate test data in suites, such tests are not possible to maintain at all. I had already rewritten the whole scan tests because of this, changing just a line of code

File: test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs [hunk 36]
```diff
@@ -0,0 +1,179 @@
...
+    // act
+    var result = sut.MatchEnv(vendor.Env);
+
+    //assert
+    result.Should().Be(expected);
+  }
+
+  [Theory]
+  [MemberData(nameof(MatchPrInput))]
+  public void MatchPrElement_ReturnsExpected(string vendorInput, IDictionary environmentInput, bool expected)
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs]: The test name and method are different

File: test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs [hunk 37]
```diff
@@ -0,0 +1,179 @@
...
+    new object[]
+    {
+      @"{}",
+      new Dictionary<string, string>(), false
+    },
+  };
+
+  [Theory]
+  [MemberData(nameof(MatchEnvInput))]
+  public void MatchEnvElement_ReturnsExpected(string vendorInput, IDictionary environmentInput, bool expected)
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/VendorMatcherTests.cs]: ditto

File: test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs [hunk 38]
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
...
+    var sut = new DefaultCiDiscovery(new Dictionary<string, string>()
+    {
+      { "GITHUB_ACTIONS", "" }, { "GITHUB_EVENT_NAME", "pull_request" }
+    });
+
     // assert
-    _sut.IsCi.Should().BeFalse();
+    sut.Server.Should().BeSameAs(CiServer.GithubActions);
+    sut.IsCi.Should().BeTrue();
+    sut.IsPr.Should().BeTrue();
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs]: You must use the partial match, we have already discussed this multiple times

File: test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs [hunk 39]
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
...
+  public void Constructor_GivenEnvironmentWithCI_CreatesInstance()
   {
+    // act
+    var sut = new DefaultCiDiscovery(new Dictionary<string, string>() { { "GITHUB_ACTIONS", "" } });
+
     // assert
-    _sut.Server.Should().BeNull();
+    sut.Server.Should().BeSameAs(CiServer.GithubActions);
+    sut.IsCi.Should().BeTrue();
+    sut.IsPr.Should().BeFalse();
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs]: ditto

File: test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs [hunk 40]
```diff
@@ -2,19 +2,42 @@ namespace SecTester.Scan.Tests.CI;
...
+  [Fact]
+  public void Constructor_GivenEmptyEnvironment_CreatesInstanceWithDefaultValues()
+  {
+    // act
+    var sut = new DefaultCiDiscovery(new Dictionary<string, string>());
+
+    // assert
+    sut.Server.Should().BeNull();
+    sut.IsCi.Should().BeFalse();
+    sut.IsPr.Should().BeFalse();
```
Comment (derevnjuk) [inline test/SecTester.Scan.Tests/CI/DefaultCiDiscoveryTests.cs]: ditto

**Review by derevnjuk — CHANGES_REQUESTED** [review]

**Review by derevnjuk — APPROVED** [review]
